x/exchange/keeper/fulfillment.go

```diff
@@ -0,0 +1,375 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	"github.com/provenance-io/provenance/x/exchange"
+)
+
+// sumAssetsAndPrice gets the sum of assets, and the sum of prices of the provided orders.
+func sumAssetsAndPrice(orders []*exchange.Order) (sdk.Coins, sdk.Coins) {
+	var totalAssets, totalPrice sdk.Coins
+	for _, order := range orders {
+		totalAssets = totalAssets.Add(order.GetAssets())
+		totalPrice = totalPrice.Add(order.GetPrice())
+	}
+	return totalAssets, totalPrice
+}
+
+// validateAcceptingOrdersAndCanUserSettle returns an error if the market isn't active or doesn't allow user settlement.
+func validateAcceptingOrdersAndCanUserSettle(store sdk.KVStore, marketID uint32) error {
+	if err := validateMarketIsAcceptingOrders(store, marketID); err != nil {
+		return err
+	}
+	if !isUserSettlementAllowed(store, marketID) {
+		return fmt.Errorf("market %d does not allow user settlement", marketID)
+	}
+	return nil
+}
+
+// FillBids settles one or more bid orders for a seller.
+func (k Keeper) FillBids(ctx sdk.Context, msg *exchange.MsgFillBidsRequest) error {
+	if err := msg.ValidateBasic(); err != nil {
+		return err
+	}
+
+	marketID := msg.MarketId
+	store := k.getStore(ctx)
+
+	if err := validateAcceptingOrdersAndCanUserSettle(store, marketID); err != nil {
+		return err
+	}
+	seller := sdk.MustAccAddressFromBech32(msg.Seller)
+	if err := k.validateUserCanCreateAsk(ctx, marketID, seller); err != nil {
+		return err
+	}
+	if err := validateCreateAskFees(store, marketID, msg.AskOrderCreationFee, msg.SellerSettlementFlatFee); err != nil {
+		return err
+	}
+
+	orders, oerrs := k.getBidOrders(store, marketID, msg.BidOrderIds, msg.Seller)
+	if oerrs != nil {
+		return oerrs
+	}
+
+	var errs []error
+	var totalSellerFee sdk.Coins
+	assetOutputs := make([]banktypes.Output, 0, len(msg.BidOrderIds))
+	priceInputs := make([]banktypes.Input, 0, len(msg.BidOrderIds))
+	addrIndex := make(map[string]int, len(msg.BidOrderIds))
+	feeInputs := make([]banktypes.Input, 0, len(msg.BidOrderIds)+1)
+	feeAddrIndex := make(map[string]int, len(msg.BidOrderIds))
+	filledOrders := make([]*exchange.FilledOrder, 0, len(msg.BidOrderIds))
+	for _, order := range orders {
+		bidOrder := order.GetBidOrder()
+		buyer := bidOrder.Buyer
+		assets := bidOrder.Assets
+		price := bidOrder.Price
+		buyerSettlementFees := bidOrder.BuyerSettlementFees
+
+		sellerRatioFee, rerr := calculateSellerSettlementRatioFee(store, marketID, price)
+		if rerr != nil {
+			errs = append(errs, fmt.Errorf("error calculating seller settlement ratio fee for order %d: %w",
+				order.OrderId, rerr))
+			continue
+		}
+		if err := k.releaseHoldOnOrder(ctx, order); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
```

Comment:
This was one of the few differences from FillAsks.

```
if sellerSettlementFlatFee != nil && !sellerSettlementFlatFee.IsZero() {
			totalSellerFee = totalSellerFee.Add(*sellerSettlementFlatFee)
		}
```

Just wanted to check there is no settlementFlatFee for the buyer?

Reply:
For the most part, you should only send funds out of an account if they've signed a TX acknowledging/accepting the loss of those funds.

The `AskOrder` struct has the `SellerSettlementFlatFee *sdk.Coin` field which is only for the flat fee. And the `BidOrder` struct has `BuyerSettlementFees sdk.Coins` which is for both the flat and ratio.

When a `BidOrder` is settled, the buyer pays their price plus the buyer settlement fees, and they receive their assets. When an `AskOrder` is settled, the seller pays the assets and receives a price (that's at least their ask price). They then pay the seller settlement fees which can come out of the price they just received.

The seller settlement flat fee can be known ahead of time, but the seller settlement ratio fee is based on the bid(s) used to settle it, so it's not known until settlement (or `FillAsks`). The seller settlement flat fee might be in a denom different from the price, but the seller ratios have to have the same price and fee denom, so the seller settlement ratio fee will always be the same denom as the price they receive. This allows us to have the sellers pay that fee out of the price that they just received. It's a gray area since they didn't actually sign a tx to have a specific fee amount removed from their account, but since they also just received the funds and (hopefully) know ahead of time that a bit of fee will come out of the price they receive.

This setup also allows someone with only `2cows` in their account (and no `chickens`) to sell them for `50chickens` while still paying the required fees for that trade. But the buyer would have to have the `50chickens` plus the required fee (flat + ratio) in their account since it's highly likely that whatever fees they're paying cannot come out of the assets they're receiving.

The `FillAsks` endpoint is similar to creating a `BidOrder` and having it immediately settled by a set of ask orders. The buyer settlement fees are paid on top of the price they're paying. So all of it must be provided in the tx, and we just need to verify that they've provided enough fee based on the total price they're paying. Each of the sellers will then get their price and pay their fees out of that.

The `FillBids` endpoint is similar to creating an `AskOrder` and having it immediately settled by a set of bid orders. Even though, with a `FillBids`, we do actually know the total settlement price, the seller ratio fee still comes out of the price they're receiving, so it's not necessary that they be indicated in the tx. So, for parity with `CreateAsk`, only the flat fee needs to be provided with a `FillBids`.

Lastly, it's possible for the seller settlement flat fee to be a denom different from the price they're receiving. When such an `AskOrder` is created, the seller must also have that flat fee in their account already, and a hold is placed on it. But if the seller settlement flat fee is the same denom as the price, we don't place a hold on it so that the `chickens` example above is still possible. During `CreateAsk`, the order is rejected if the flat fee denom is the same as the price, and flat fee + ratio fee at the ask price is more than the ask price. E.g. if the seller flat fee is `10nhash` and their price is `12nhash`, and the ratio fee ends up being `3nhash`, settlement would result in the seller receiving `12nhash`, then having to pay `13nhash` for the pleasure. So we prevent that.